DataBiosphere/leonardo
# [IA-4268] RStudio on GKE as a new App Type

__Jira ticket__: https://broadworkbench.atlassian.net/browse/[ticket_number]



This PR implements the installation of the new RStudio helm chart developed by AOU: https://github.com/DataBiosphere/terra-app/pull/34

Sister PR in terra helmfile repo to disable this App in prod until we successfully go through the production readiness checklist https://github.com/broadinstitute/terra-helmfile/pull/4195

## Summary of changes



### What

- Creates a new RStudio app type that supports welder file syncing and detachable PDs
- Add this app type to the app metrics we monitor
- When creating the app in the GKE interpreter, the specificity of RStudio is the need to create a staging bucket for Welder to use (to dump logs into)

### Why

- AOU needs RStudio to run as an app in Q2 2023, which means that we unfortunately do not have the time to enable new app types via a custom helm chart URL option, but have to create a new app type from scratch (similar to the work done for Hail and WDS recently)

## Testing these changes

### What to test

- Update your BEE with the latest image generated by this PR
- Make sure that you have a google project ready to use
- From your BEE leo swagger page, create an RStudio app using the apps/createApp v1 endpoint
```
{
  "diskConfig": {
    "name": "disk1"
  },
  "appType": "RSTUDIO",
  "customEnvironmentVariables": {
    "WORKSPACE_NAME": "<YOUR_WORKSPACE_NAME>"
  }
}
```
- Monitor your leonardo BEE log, as well as the k8s workload to ensure that the app is being installed properly and both the rstudio and welder pods are running
- To set up Welder you need to:
- - Create a storage bucket in your google project (e.g. `test-welder`) and make sure to give it the correct permissions (easier to copy the permissions  from the staging bucket that was created with the app in that google project)
- - Send a welder post request to create a storage link to that test bucket, e.g.
```
curl 'https://leonardo.<YOUR-BEE-NAME>.bee.envs-terra.bio/proxy/google/v1/apps/<YOUR-GOOGLE-PROJECT>/<YOUR-APP-NAME>/welder-service/storageLinks' \
  -H 'authority: leonardo.<YOUR-BEE-NAME>.bee.envs-terra.bio' \
  -H 'accept: */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'authorization: Bearer <YOUR-ACCESS-TOKEN>' \
  -H 'content-type: application/json' \
  -H 'origin: https://terra.<YOUR-BEE-NAME>.bee.envs-terra.bio' \
  -H 'referer: https://leonardo.dsde-dev.broadinstitute.org' \
  --data-raw '{"localBaseDirectory":"","localSafeModeBaseDirectory":"","cloudStorageDirectory":"gs://<YOUR-BUCKET-NAME>/notebooks","pattern":".+(\\.R|\\.Rmd)$"}' \
  --compressed
```
- Generate a cookie in your browser from the leo BEE swagger page (proxy/setCookie endpoint)
- You might need to use something like the ModHeader pluggin to add `https://leonardo.dsde-dev.broadinstitute.org` as the Referer
- Open the RStudio app (the url should be `https://leonardo.e<YOUR-BEE-NAME>.bee.envs-terra.bio/proxy/google/v1/apps/<YOUR-GOOGLE-PROJECT>/<YOUR-APP-NAME>/rstudio-service/`). Note that the first time it might redirect you to get authenticated, I have not quite figured this one yet. So what I ended up doing is to first go to the `/rstudio-service/auth-sign-in` endpoint that redirects me to the swagger page and then I can access the url above.
- You should be able to create R and Rmd files, and you can then check that they are being synced by welder to your test storage bucket
- Also load a non R file, like a picture of a bunny 🐰 so we can test if the persistent disk reattachment works in the steps below
- Delete the app but keep the disk
- Create another RStudio app using the same disk
- Open the app again and make sure the data from the old app is still in there



### Who tested and where

  - _NB: Rerun automation tests on this PR by commenting `jenkins retest` or `jenkins multi-test`._
